File: Daily/PY/Leetcode224-基本计算器.py
class Solution:
    def calculate(self, s: str) -> int:
        n = len(s)
        ret = 0
        ops = [1]
        sign = 1
        i = 0
        while i < n:
            if s[i] == ' ':
                i += 1
            elif s[i] == '+':
                sign = ops[-1]
                i += 1
            elif s[i] == '-':
                sign = -ops[-1]
                i += 1
            elif s[i] == '(':
                ops.append(sign)
                i += 1
            elif s[i] == ')':
                ops.pop()
                i += 1
            else:
                sum = 0
                while i < n and s[i] >= '0' and s[i] <= '9':
                    sum = sum * 10 + ord(s[i]) - ord('0')
                    i += 1
                ret += sign * sum
        return ret

File: Daily/sbgo/Leetcode224-基本计算器.go
package sbgo

func calculate(s string) int {
	n := len(s)
	i := 0
	ops := []int{1}
	ret := 0
	sign := 1
	for i < n {
		if s[i] == ' ' {
			i++
		} else if s[i] == '+' {
			sign = ops[len(ops)-1]
			i++
		} else if s[i] == '-' {
			sign = -ops[len(ops)-1]
			i++
		} else if s[i] == '(' {
			ops = append(ops, sign)
			i++
		} else if s[i] == ')' {
			ops = ops[:len(ops)-1]
			i++
		} else {
			sum := 0
			for i < n && s[i] >= '0' && s[i] <= '9' {
				sum = sum*10 + int(s[i]-'0')
				i++
			}
			ret += sum * sign
		}
	}
	return ret
}
